SU2_CFD/include/output/COutput.hpp

```diff
@@ -0,0 +1,774 @@
+   * \param[in] iMarker - The index of the marker.
+   */
+  inline void SetHistoryOutputPerSurfaceValue(string name, su2double value, unsigned short iMarker){
+    if (historyOutputPerSurface_Map.count(name) > 0){
+      historyOutputPerSurface_Map[name][iMarker].value = value;
+    } else {
+      SU2_MPI::Error(string("Cannot find output field with name ") + name, CURRENT_FUNCTION);
+    }
+  }
+  
+  /*!
+   * \brief Add a new field to the volume output.
+   * \param[in] name - Name for referencing it (in the config file and in the code).
+   * \param[in] field_name - Header that is printed in the output files.
+   * \param[in] groupname - The name of the group this field belongs to.
+   * \param[in] description - Description of the volume field.
+   */
+  inline void AddVolumeOutput(string name, string field_name, string groupname, string description){
+    volumeOutput_Map[name] = VolumeOutputField(field_name, -1, groupname, description);
+    volumeOutput_List.push_back(name);
+  }
+  
+  
+  /*!
+   * \brief Set the value of a volume output field
+   * \param[in] name - Name of the field.
+   * \param[in] value - The new value of this field.
+   */
+  su2double GetVolumeOutputValue(string name, unsigned long iPoint);
+  
+  /*!
+   * \brief Set the value of a volume output field
+   * \param[in] name - Name of the field.
+   * \param[in] value - The new value of this field.
+   */
+  void SetVolumeOutputValue(string name, unsigned long iPoint, su2double value);
+
+  /*!
+   * \brief Set the value of a volume output field
+   * \param[in] name - Name of the field.
+   * \param[in] value - The new value of this field.
+   */
+  void SetAvgVolumeOutputValue(string name, unsigned long iPoint, su2double value);
+  
+  /*!
+   * \brief CheckHistoryOutput
+   */
+  void CheckHistoryOutput();
+  
+  /*!
+   * \brief Open the history file and write the header.
+   * \param[in] config - Definition of the particular problem.
+   */
+  void PrepareHistoryFile(CConfig *config);
+
+  /*!
+   * \brief Load up the values of the requested volume fields into ::Local_Data array. 
+   * \param[in] config - Definition of the particular problem.
+   * \param[in] geometry - Geometrical definition of the problem.
+   * \param[in] solver_container - The container holding all solution data.
+   */
+  void LoadDataIntoSorter(CConfig* config, CGeometry* geometry, CSolver** solver);
+      
+  /*!
+   * \brief Postprocess_HistoryData
+   * \param[in] config - Definition of the particular problem.
+   */
+  void Postprocess_HistoryData(CConfig *config);
+
+  /*!
+   * \brief Postprocess_HistoryFields
+   * \param[in] config - Definition of the particular problem.
+   */
+  void Postprocess_HistoryFields(CConfig *config);
+  
+  /*!
+   * \brief Check whether we should print output.
+   * \param[in] iIter - Current iteration.
+   * \param[in] iFreq - Frequency of output printing.
+   */
+  inline bool PrintOutput(unsigned long iIter, unsigned long iFreq) {
```

Comment:
Is it possible to write the initial condition with the new output structure (before first iteration)? This is something I have been meaning to add, since it is very useful when setting up and debugging custom problems

Reply:
This probably needs an extra call of the output routines, because right now they are always called at the end of the iteration. Maybe we can do it when you set the number of iterations to 0 ?

Reply:
Actually it already works like that. Just set the `INNER_ITER`/`ITER` to 0.